### PR TITLE
Define native Client interface for AsymmetricSign

### LINKED ISSUE #10: Implement KmsRsaKey for performing signing operations RSA keys
`KmsRsaKey` contains a Cloud HSM resource ID and methods for performing operations with that key.

- [x] Define `RsaKey` interface (should be based on / copied from `RSA_METHOD` from OpenSSL). While signing operations are the main focus of this ticket, we might as well go ahead and define the whole interface at this point.
- [x] Define `ApiClient` interface with minimum set of KMS operations needed to implement `KmsRsaKey`
- [x] Implement `KmsRsaKey` methods for `RsaKey` interface
- [x] Write unit tests for `KmsRsaKey` by creating mock `ApiClient`

### LINKED ISSUE #11: Implement minimal GrpcApiClient wrapper around protobuf generated code
`GrpcApiClient` is a wrapper for the [`google.cloud.kms.v1` RPC functions](https://cloud.google.com/kms/docs/reference/rpc/google.cloud.kms.v1). The purpose of wrapping this class around the gRPC client is to allow for authenticated credentials to be associated with the client as well as provide a interface for injecting mock `ApiClient`s for testing purposes.

- [x] Implement `GrpcApiClient` methods for minimal `ApiClient` interface (essentially just wrappers around protobuf generated code)
- [x] Implement `GrpcApiClient` credential instantiation via [`grpc::GoogleDefaultCredentials`](https://grpc.io/docs/guides/auth/#using-google-token-based-authentication)

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Part of #10 and #11.

Defines a minimal `Client` interface for making `AsymmetricSign` requests against the Cloud KMS API, along with several native definitions of request structures and resources.

### CODE REVIEW

File: src/backing/client/client.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#include "src/backing/client/asymmetric_sign_response.h"
+
+namespace kmsengine {
+namespace backing {
+namespace client {
+
+// Defines the interface used to communicate with the Google Cloud KMS API.
+class Client {
+ public:
+  virtual ~Client() = 0;
```
Comment (ZacharyEspiritu) [inline src/backing/client/client.h]: This needs to be:

```c++
virtual ~Client() = default;
```

**Review by ajfeldma-google — COMMENTED** [review]
These classes seem to essentially replicate the proto definitions. Why can't we just use the protos?

**Comment by ZacharyEspiritu** [comment]
It separates the "gRPC-ness" of the eventual `GrpcClient` from the `Client` interface that the rest of the codebase sees. This is useful if someone wants to add another type of `Client` in the future, and also is similar to how some of the existing Google Cloud C++ client libraries are structured (see [`RawClient` interface](https://github.com/googleapis/google-cloud-cpp/blob/master/google/cloud/storage/internal/raw_client.h) from the Storage library).

**Comment by ZacharyEspiritu** [comment]
One other thing I forgot to mention---defining the classes like this would allow the user of the `Client` interface to ignore things like the `crc32c` checksum field once that is officially added to the protobuf definitions. (See `ciphertext_crc32c` in [the API](https://cloud.google.com/kms/docs/reference/rpc/google.cloud.kms.v1#asymmetricdecryptrequest)---this is a beta field and not currently in the canonical protobuf definitions.) Once this field is added, the engine can just add a getter to the native classes that does something like:

```c++
std::string get_ciphertext_crc32c() {
  return Crc32c::crc32c(ciphertext);
}
```

And the calling code wouldn't have to worry about how `crc32c` is actually defined (it would be up to `Client` implementations to handle it). (I would have set this up already, but it looks like its not in the official protobuf definitions from [googleapis/googleapis](https://github.com/googleapis/googleapis/tree/master/google/cloud/kms/v1).)

**Review by tonyliaoss — APPROVED** [review]

**Comment by tonyliaoss** [comment]
> These classes seem to essentially replicate the proto definitions. Why can't we just use the protos?

I do think that using protos as the source of truth is neater, but if the goal is to make the engine layer agnostic to gRPC, then I can't really think of a cleaner way than custom defining C++ request/responses. So LGTM for now, for me.

**Comment by ZacharyEspiritu** [comment]
After thinking about this for a week, I can't think of a case where one wouldn't want to use gRPC and so maybe this approach is not needed. The `RsaKey` abstraction separates the bridge layer from the gRPC client and so the bridge never would have to refer to gRPC constructs (with the exception of the `DigestType` enum---but I can just expose the gRPC `DigestType` as an alias).

**Comment by ajfeldma-google** [comment]
> After thinking about this for a week, I can't think of a case where one wouldn't want to use gRPC and so maybe this approach is not needed. The `RsaKey` abstraction separates the bridge layer from the gRPC client and so the bridge never would have to refer to gRPC constructs (with the exception of the `DigestType` enum---but I can just expose the gRPC `DigestType` as an alias).

Make sense to me. I don't think you even need to alias the KMS protos in the `backing` library since it's pretty tied to KMS.

**Comment by ZacharyEspiritu** [comment]
Closed in favor of #54 and follow-up PRs.

**Pull request closed without merging**